### ISSUE TITLE
Sleeping and Thirst

### ISSUE THREAD

**Issue by PitchBright** [thetorine/thirstmod#7]
These are the things that people on my server are noticing. I can gather more info if you require. It would be nice to be able to configure the amount of thirst lost by sleeping.
# 

BUG description: After you sleep, your drink completely drains and you are given nausea and start dying.. If you have no drinks on you, you will die. We think it’s because we can now get into bed before the clock is pointing completely straight up. Not sure if it’s an issue that can be fixed, but at the same time, I know a lot of people spam their beds before getting in.
Found by: Dutty
imgur link to screencapture: -

^^ To be added to Dutty’s
BUG description: When sleeping just past the 90 degree angle going into the night on the clock, you only lose 2.5 water droplets. After logging on while it being the middle of the night and sleeping, I woke up with only 1 water droplet missing. tl;dr It seems that the later into the night you sleep, the less water droplets you lose but not as significant as the full bar being depleted. The complete depletion of the hydration bar might actually be a bug instead of just a case of “realism.”
Found by: ger_
imgur link to screencapture: 

**
BUG Description: I had 7 drops of water and when i clicked to sleep I got the message ‘to thirsty to sleep’?
Found by: Ahz
imgur link: http://imgur.com/eVVFALj

**Comment by thetorine** [comment]
Currently the player loses half a droplet of thirst every 2000 ticks they sleep for. (The maximum of ticks the player can sleep for I believe is around 11000 ticks) If the mod determines that the thirst when the player wakes up will be less than or equal to four droplets, then sleep is prevented with the message "You are too thirsty to sleep.". 

The current method for calculating thirst loss and determining whether to allow the player to sleep is as follows:
1. Player attempts to sleep in bed.
2. Mod calculates thirst loss using the formula: 
    loss = (daylength - currenttime)/2000
3. If the amount of thirst to lose places the thirst bar at 4 droplets or less, then sleep is inhibited. Otherwise sleep is permitted and the thirst loss is removed upon waking up.

According to these bug reports, the mod is failing to recognise that sleeping will drain the thirst bar below the minimum of four droplets, thus allowing the player to sleep, resulting in too much thirst loss. This also seems to be true for the opposite where it seems to think the player's thirst is too low and preventing the player from sleeping. 

The most plausible error that could cause this is that the syncing of the data from server to client is malfunctioning. Most likely due to an error in the mod itself and nothing to do with your setup.  I'll look into the issue, try to reproduce it and fix it. 

Thanks for letting me know.

**Comment by PitchBright** [comment]
If you need me or my players to test or produce any other scenarios that will assist you in collecting information for this, just let me know.

Also, if you can make the features of this function configurable (ie. Amount and/or Rate of thirst loss while asleep, effects, and effect level, and effect duration of waking up thirsty)… it would be GREATLY appreciated. To have that kind of fine-grain control would go a long way for various users of the mod who wish to customize their experience, in varying environments. (For example, all of my people go to bed, simultaneously, EVERY night the moment the sun hits the horizon, as we play in Survival mode. The effects of waking up thirsty may be too steep for our experience, where other servers may not feel it is a problem.)

**Comment by thetorine** [comment]
If you can kindly provide a list of anything in the mod that should be configurable, I will surely add it.

**Comment by PitchBright** [comment]
I think for this particular instance… the configurables that would really benefit all users would be:

-Effect enabled/disabled (true or false)
-Effect ID (if effect is enabled)
-Effect Duration (if effect is enabled)
-Effect Level (if effect is enabled)
-Thirst Loss during sleep multiplier (ie.  0 = no loss, 1 = 1x Standing Still loss, 1.5 = 1.5x Standing Still loss)

Other stuff may come to my mind, in the future. I can start a new issue(feature request) for those at that time though.

**Comment by thetorine** [comment]
What do you mean by effect? Is it referring to the loss of thirst when sleeping?

**Comment by PitchBright** [comment]
The "effect" you have now, when waking up thirsty, is Nausea II

**Comment by thetorine** [comment]
Yeah, nausea shouldn't occur when waking up because the player shouldn't be able to sleep if their thirst will drop to four droplets or lower. This is a bug and I'm looking into a fix. I just haven't been able to reproduce it on single player or multiplayer.

**Comment by PitchBright** [comment]
Try clicking on a bed, to get into it… a few seconds BEFORE the minecraft clock shows the "night" reaching the very top of the clock. If you get into a bed when the Day/Night line on the clock is at the exact top of the clock, then it doesn't happen… but since MC let's you get into bed just a few seconds before that point… you can get into bed "early" (relative to MC 1.4.6).

**Comment by thetorine** [comment]
I've tried spam clicking the right mouse button to get into bed just seconds prior to the point you described but I still am not getting any error or losing too much thirst when I sleep. Are you sure this occurs on the newest version because I changed the way night was detected?

I'll add a method to disable thirst loss when asleep but will work on adding other configurable options later. School is preventing me from working on this mod full time. Sorry about that.

**Comment by PitchBright** [comment]
I'll re-test it for ya. I think this bug report was from the version before last, and I think I remember thinking that you fixed it in the most recent version. I'll test for it on 1.8.14

**Comment by thetorine** [comment]
Alright, thanks mate.

**Comment by PitchBright** [comment]
The issue seems to be gone!
